### DISCUSSION TITLE
Meaning of Tree.sleep()

### DISCUSSION THREAD

**Discussion by MatteoAlbi** [BehaviorTree/BehaviorTree.CPP#656] [Q&A] [answered]
My question is why do we need to add a sleep interval in the tick cycle of the tree?

**Comment by MatteoAlbi** [comment, marked as answer]
Answer from @facontidavide :

Since the concept of ticking a tree is basically "polling", we don't want to consume 100% of the cpu, bombarding an asynchronous action with ticks.

The smart thing we do is that the proposed sleep is actually a std::condition_variable in disguise.

So, contrariwise to a regular sleep, it can be interrupted if a wake up signal is sent

**Comment by MatteoAlbi** [comment]
And in which cases a wake up signal is sent? Moreover, what could actually happen by shortening the sleep period?

**Reply by facontidavide** [reply]
A well implemented asynchornous action with its own thread, should invoke `emitWakeUpSignal` when it is ready.

For instance, imagine that you asynchronous Action is waiting for  a reply from a server (in its own thread).
As soon as this reply is reached, you should call `emitWakeUpSignal` to notify that you are ready to be complete your action in the next `tick`.
